docs/proposals/20220707-service-primitive.md

```diff
@@ -0,0 +1,87 @@
+# Service Expose Support
+Authors:
+- [nullday](https://github.com/aseaday)
+
+## Summary
+
+This proposal is designed for exposing service from the environment running application such as Juypter for external use. It introduce a first-level function `serve` and its domain related packages:
+
+- host
+- cloud
+
+And maybe a utils package for the support of define the default service port or socket predefined:
+
+- presets
+
+## Motivation
+
+There a lot of AI/ML develop toolkits could be accessible from Web or GUI Client.  Users need a interface to configure the service which enviroment runtime could expose. The juypter is a apparent case need to be supported.
+
+There is also such the kind of demand when users run their env in the cloud which accessing the service could be a little complex depends on the cloud conditions. For example, ops team allows exposing the service or ingress or only permit a port-forward.
+
+## Goals
+- http/tcp support
+- local support:
+    - linux/mac local access
+    - wsl access from the windows
+- remote support:
+    - k8s port-foward support
+    - k8s service/node-port support for http
+
+Note: There is no need for support non-tcp protocl for now.
+
+## Proposal
+
+Add `def serve` as the first level function as `def build` but could be omitted.
+
+```python
+def build():
+    base(os="ubuntu20.04", language="python")
+    # Use `config.jupyter(password="")` 
+    # if you do not need to set up password.
+    config.jupyter(password="password")
+
+def serve():
+    pass
+```
+
+import two package `host` and `cloud` in the use of `server` function. The follow example give a user story user want to export juypter port to the host port 1234:
+
+```python
+def build():
+    base(os="ubuntu20.04", language="python")
+    # Use `config.jupyter(password="")` 
+    # if you do not need to set up password.
+    config.jupyter(password="password")
+def serve():
+    host.bind(env_port=presets.juypter.port,host_port=1234) # The bind should judge the wsl automatically
+```
+
+for the cloud user scenario:
+
+
+```python
+def build():
+    base(os="ubuntu20.04", language="python")
+    # Use `config.jupyter(password="")` 
+    # if you do not need to set up password.
+    config.jupyter(password="password")
+def serve():
+    cloud.k8s.bind(env_port=presets.juypter.port,host_port=1234) # It would use port forward
```

Comment:
It relates to #303 

I am not sure if we should introduce the `cloud` package here. But I do think we need a mechanism like `host.bind` to support other services like #527 TensorBoard.

Reply:
* The name `serve` has ambiguity with `serving` https://github.com/tensorchord/envd/issues/157

Reply:
> * The name `serve` has ambiguity with `serving` [discussion(lang): Should we support serving? #157](https://github.com/tensorchord/envd/issues/157)

I think `expose` is a more appropriate keyword.